swiftpkg/internal/registry_swift_package.bzl

```diff
@@ -0,0 +1,215 @@
+_PACKAGE_SWIFT_NAME = "Package.swift"
+
+# The headers to use when making requests to the registry.
+_REGISTRY_JSON_ACCEPT_HEADER = {
+    "Accept": "application/vnd.swift.registry.v1+json",
+}
+_REGISTRY_SWIFT_ACCEPT_HEADER = {
+    "Accept": "application/vnd.swift.registry.v1+swift",
+}
+_REGISTRY_ZIP_ACCEPT_HEADER = {
+    "Accept": "application/vnd.swift.registry.v1+zip",
+}
+
+def _get_id(id):
+    """Returns the scope and name from the package identifier.
+
+    https://github.com/swiftlang/swift-evolution/blob/main/proposals/0292-package-registry-service.md#package-identity
+    """
+    scope, name = id.split(".")
+    return struct(
+        scope = scope,
+        name = name,
+    )
+
+def _get_registries(registries_json):
+    """Returns the registry structs from the registries JSON."""
+    default_registry = None
+    scoped_registries = {}
+
+    for registry_key, registry in registries_json.items():
+        if registry_key == _DEFAULT_REGISTRY_KEY:
+            default_registry = registry
+        else:
+            scoped_registries[registry_key] = registry.get("url")
+
+    if not default_registry and not scoped_registries:
+        fail("No registries were parsed from the `registries` field: {registries}".format(registries = registries_json))
+
+    return struct(
+        default = default_registry.get("url"),
+        scoped = scoped_registries,
+    )
+
+def _get_registry_url(*, id, registries):
+    """Returns the registry URL for the given package identifier, preferring a scoped registry if it is defined."""
+    return registries.scoped.get(id.scope, registries.default)
+
+def _download_and_parse_package_json(*, id, output, repository_ctx, registry_url, version):
+    """Downloads and parses the package JSON metadata from the registry.
+
+    https://github.com/swiftlang/swift-package-manager/blob/main/Documentation/PackageRegistry/Registry.md#42-fetch-information-about-a-package-release
+    """
+    package_json_url = paths.join(
+        registry_url,
+        id.scope,
+        id.name,
+        version,
+    )
+    repository_ctx.download(
+        headers = _REGISTRY_JSON_ACCEPT_HEADER,
+        url = package_json_url,
+        output = output,
+    )
+
+    return json.decode(repository_ctx.read(output))
+
+def _download_archive(*, checksum, id, output, repository_ctx, registry_url, version):
+    """Downloads the specific version of the archive from the registry.
+
+    https://github.com/swiftlang/swift-package-manager/blob/main/Documentation/PackageRegistry/Registry.md#44-download-source-archive
+    """
+    archive_url = paths.join(
+        registry_url,
+        id.scope,
+        id.name,
+        "{version}.zip".format(version = version),
+    )
+    repository_ctx.download_and_extract(
+        headers = _REGISTRY_ZIP_ACCEPT_HEADER,
+        url = archive_url,
+        output = output,
+        sha256 = checksum,
+        stripPrefix = "{scope}.{name}".format(scope = id.scope, name = id.name),
```

Comment:
Oh rough yeah, I had been using `<name>-<version>` for the prefix in my fake registry but then our Artifactory implementation was using `scope.name`. 

I'll update this to work like you described since it looks like we cant use `strip_prefix` here

Reply:
Updated this! Used a similar approach except using `cp` and `rm` instead of `find`